pageserver/src/tenant/storage_layer/layer.rs

```diff
@@ -0,0 +1,1150 @@
+                        this.desc.file_size,
+                    ),
+                )
+                .await;
+
+                let result = match result {
+                    Ok(size) => {
+                        timeline.metrics.resident_physical_size_gauge.add(size);
+                        Ok(())
+                    }
+                    Err(e) => {
+                        Err(e)
+                    }
+                };
+
+                if let Err(res) = tx.send(result) {
+                    match res {
+                        Ok(()) => {
+                            // our caller is cancellation safe so this is fine; if someone
+                            // else requests the layer, they'll find it already downloaded
+                            // or redownload.
+                            //
+                            // however, could be that we should consider marking the layer
+                            // for eviction? alas, cannot: because only DownloadedLayer
+                            // will handle that.
+                        },
+                        Err(e) => {
+                            // our caller is cancellation safe, but we might be racing with
+                            // another attempt to initialize. before we have cancellation
+                            // token support: these attempts should converge regardless of
+                            // their completion order.
+                            tracing::error!("layer file download failed, and additionally failed to communicate this to caller: {e:?}");
+                        }
+                    }
+                }
+
+                Ok(())
+            }
+            .in_current_span(),
+        );
+        match rx.await {
+            Ok(Ok(())) => {
+                if let Some(reason) = self
+                    .needs_download()
+                    .await
+                    .map_err(DownloadError::PostStatFailed)?
+                {
+                    // this is really a bug in needs_download or remote timeline client
+                    panic!("post-condition failed: needs_download returned {reason:?}");
+                }
+
+                self.consecutive_failures.store(0, Ordering::Relaxed);
+
+                Ok(())
+            }
+            Ok(Err(e)) => {
+                let consecutive_failures =
+                    self.consecutive_failures.fetch_add(1, Ordering::Relaxed);
+                tracing::error!(consecutive_failures, "layer file download failed: {e:#}");
+                let backoff = utils::backoff::exponential_backoff_duration_seconds(
+                    consecutive_failures.min(u32::MAX as usize) as u32,
+                    1.5,
+                    60.0,
+                );
+                let backoff = std::time::Duration::from_secs_f64(backoff);
+
+                // unless we get cancelled, we will hold off the init semaphore
+                tokio::time::sleep(backoff).await;
```

Comment:
Maybe my attention is dwindling but it's not clear to me what this "init semaphore" is.

And did you really mean "holding off" (if yes, what does it mean) vs "holding on to the permit"?
My confusion here is likely from not understanding what "init semaphore" is referred to in the comment.

Reply:
> Unlike tokio::sync::OnceCell, there's a semaphore for which you get OwnedSemaphorePermit. With those, you can unlock the mutex which is around the "OnceCell" for the duration of the download, enabling evict_and_wait to inspect the same mutex (using LayerInner::get) without waiting for downloads.

That is a nice, concise explanation of what that API does.
Can you copy-paste the above as a doc comment on the respective heave_once_cell API?

> By that I mean the heavier_once_cell::OnceCell semaphore which is used for initialization.
> ...
> So this backoff hits on the line let value = factory().await?;, hopefully stalling the next attempt to download and initialize.

Let me turn that into a suggestion that would have been more helpful to me than the current comment. Alternatively, I think removing the comment would also do a job. The coalescing behavior is what one would expect in the first place with a OnceCell


```suggestion
                // This function is part of the `download` factory future that `get_or_maybe_init` passes to `heavier_once_cell::OnceCell::get_or_init`.
                // The `get_or_init` coalesces concurrent callers to the first one.
                // So, if we back-off here, concurrent callers that need the layer will
                // coalesce on our back-off. That's the desired behavior, there should
                // be no early retry just because there's increased interest.
                tokio::time::sleep(backoff).await;
```

Reply:
> > Unlike tokio::sync::OnceCell, there's a semaphore for which you get OwnedSemaphorePermit. With those, you can unlock the mutex which is around the "OnceCell" for the duration of the download, enabling evict_and_wait to inspect the same mutex (using LayerInner::get) without waiting for downloads.
> 
> That is a nice, concise explanation of what that API does. Can you copy-paste the above as a doc comment on the respective heave_once_cell API?

Tried to move, I think the first sentence already covers it:

```
/// Custom design like [`tokio::sync::OnceCell`] but using [`OwnedSemaphorePermit`] instead of
/// `SemaphorePermit`, allowing use of `take` which does not require holding an outer mutex guard
/// for the duration of initialization.
```

Comment on the backoff removed.